File: app/src/main/java/io/github/howiezuo/unsplash/feature/main/MainViewModelModule.java
package io.github.howiezuo.unsplash.feature.main;

import android.support.v7.app.AppCompatActivity;

import dagger.Module;
import dagger.Provides;
import io.github.howiezuo.unsplash.app.ActivityScope;
import io.github.howiezuo.unsplash.app.ViewModelHolder;
import io.github.howiezuo.unsplash.util.ActivityUtils;

@Module
public class MainViewModelModule {

    private final AppCompatActivity activity;
    private String tag;

    public MainViewModelModule(AppCompatActivity activity, String tag) {
        this.activity = activity;
        this.tag = tag;
    }

    @Provides
    @ActivityScope
    MainViewModel provideMainViewModel() {
        ViewModelHolder<MainViewModel> holder = (ViewModelHolder<MainViewModel>) activity.getSupportFragmentManager().findFragmentByTag(tag);

        if (holder != null && holder.getViewModel() != null) {
            return holder.getViewModel();
        } else {
            MainViewModel viewModel = new MainViewModel();
            ActivityUtils.addFragmentToActivity(activity.getSupportFragmentManager(), ViewModelHolder.createHolder(viewModel), tag);
            return viewModel;
        }
    }
}

File: app/src/main/java/io/github/howiezuo/unsplash/feature/main/MainViewModelModule.kt
package io.github.howiezuo.unsplash.feature.main

import android.support.v7.app.AppCompatActivity
import dagger.Module
import dagger.Provides
import io.github.howiezuo.unsplash.app.ActivityScope
import io.github.howiezuo.unsplash.app.ViewModelHolder
import io.github.howiezuo.unsplash.util.ActivityUtils


@Module
class MainViewModelModule(
        private val activity: AppCompatActivity,
        private val tag: String) {

    @Provides
    @ActivityScope
    fun provideMainViewModel(): MainViewModel {
        val holder = activity.supportFragmentManager.findFragmentByTag(tag) as? ViewModelHolder<MainViewModel>
        if (holder != null && holder.viewModel != null) {
            return holder.viewModel!!
        } else {
            val vm = MainViewModel()
            ActivityUtils.addFragmentToActivity(activity.supportFragmentManager, ViewModelHolder.createHolder(vm), tag)
            return vm
        }
    }

}